jgrapht-core/src/main/java/org/jgrapht/alg/flow/EdmondsKarpMaximumFlow.java

```diff
@@ -59,20 +59,29 @@
  * <p>For more details see Andrew V. Goldberg's <i>Combinatorial Optimization
  * (Lecture Notes)</i>.
  *
+ * Note: even though the algorithm accepts any kind of graph, currently only Simple directed and undirected graphs are supported (and tested!).
+ *
+ *
+ *
  * @author Ilya Razensteyn
  */
+
+/* * JK: Issues with the current implementation:
+ * 1. The internal data structures are completed rebuild each time the algorithm is invoked. Even if the
+ * graph and edge capacities don't change, but the source/sink pair change, the data structures get recomputed.
+ * 2. The algorithm uses custom data structures to represent a flow network internally. This however could easily be
+ * replaced by a DirectedWeightedGraph. However, due to some overhead in the graphs prior to version 0.9.3, it turned out
+ * beneficial to use the custom structures instead. In the next iteration we should replace them by proper graph structures.
+ * This will improve robustness and readability significantly.*/
 public final class EdmondsKarpMaximumFlow<V, E>
     extends MaximumFlowAlgorithmBase<V, E>
 {
-    private DirectedGraph<V, E> network; // our network
-
-    private double epsilon; // tolerance (DEFAULT_EPSILON or user-defined)
 
     private VertexExtension currentSource; // current source vertex
     private VertexExtension currentSink; // current sink vertex
 
     private final ExtensionFactory<VertexExtension> vertexExtensionsFactory;
-    private final ExtensionFactory<EdgeExtension> edgeExtensionsFactory;
+    private final ExtensionFactory<AnnotatedFlowEdge> edgeExtensionsFactory;
```

Comment:
In the original code there was an Extension, which was implemented or extended by a base extension which was then extended by ExtendedEdge. Then there were functions such as protected EdgeExtension extendedEdge(E e){return edgeExtended(e)}.There was a lot of something-extends-something which made this simple algorith rather complex. As you've noticed, function mentioned above has been eliminated and code has been simplified. If you know a catchy name for VertexExtension then I'm happy to replace that name as well.